Changed private message" button to use <a href> to open Private Messa…

tsukimi2 opened the pull request:
Fixes #11910

Changed private message" button to use <a href to open Private Message page in same tab with normal mouse click, and open in a new tab with right mouse click mouse context menu.  Pass the selected user ID in profile component as part of url query when clicking the private message icon and opening the PMs page.

tsukimi2 committed:
Changed private message" button to use <a href> to open Private Message page in same tab.

Changed private message" button to use <a href> to open Private Message page in same tab with normal mouse click, and open in a new tab with right mouse click mouse context menu.

paglias commented on website/client/src/components/userMenu/profile.vue:
```diff
@@ -10,10 +10,12 @@
-            class="svg-icon message-icon"
-            v-html="icons.message"
-          ></div>
+          <a :href="'/private-messages?uuid=' + user._id">
```
can we use a `router-link` component? It'll be rendered as an `a` tag but meant to work with Vue router links https://router.vuejs.org/api/#router-link-props

paglias commented on website/client/src/pages/private-messages.vue:
```diff
@@ -630,7 +630,16 @@ export default {
     await this.reload();
 
-    const data = this.$store.state.privateMessageOptions;
+    let data = this.$store.state.privateMessageOptions;
```
I'd say this is the perfect place to use Navigation Guards https://router.vuejs.org/guide/advanced/navigation-guards.html#in-component-guards, they'll be called when the route is loaded

tsukimi2 committed:
Merge remote-tracking branch 'upstream/develop' into bugfix-pm-obscured

tsukimi2 committed:
Use router-link and navigation guard as requested.

Change from <a href to router-link in profile.vue and use navigation guard to obtain selected user data from server instead of doing so in mount method as requested.

paglias commented:
@tsukimi2 sorry but I'm still seeing the issue with the current code, the members modal remains open in the PM pages

tsukimi2 commented:
> the members modal remains open in the PM pages

@paglias By "the members modal remains open in the PM pages", do you mean members modal remains open when the PM page is opened in an existing tab (i.e. when mouse left-click on the message icon), or do you mean members modal remains open on the original tab when the PM page is being opened in a new tab (i.e. the case when mouse right-click and then choosing "open in a new tab")?

paglias commented:
@tsukimi2 in an existing tab

tsukimi2 committed:
Added habitica::dismiss-modal to hide members modal

tsukimi2 committed:
Merge remote-tracking branch 'upstream/develop' into bugfix-pm-obscured

tsukimi2 committed:
Suggestion from lint

tsukimi2 commented:
@paglias Sorry, tested the previous code on mac and seemed to work properly there but did not realize that for some reason it's not working on linux when testing the workflow more than once.

Anyway, the updated code was manually tested on chrome and safari on mac, and chrome and firefox on linux.  But I'm not entirely sure whether this is the correct way to solve the problem.  Under further testing, I found that when I use the code (in line 654 on private-messages.vue)

this.$root.$emit('bv:hide::modal', 'members-modal');

alone to close the members modal, it works only on mac but not on linux.  However, when I use the code

this.$root.$emit('habitica::dismiss-modal', 'members-modal');

alone it works on linux and not on mac.  So for now I put both lines I put both lines into the modified code and seems to work on both mac and linux (have not tried out on windows since I have not set up the habitica development environment on windows), though I do not understand why this works.  Feeling a bit uneasy about this addition since it feels a bit hack-ish and I do not have the dev environment to test it out on windows yet.

I also tried to observe "this.$store.state.modalStack" (line 696) when running the code, which is the part of the codebase where it actually hides modals.
https://github.com/HabitRPG/habitica/blob/8efed3724150628ace963e4ca2a0554814a73ecf/website/client/src/app.vue#L583-L612

the code seems to be able to clear off the this.$store.state.modalStack with length = 0 after clicking the private messages icon on the profile.vue, so not entirely sure myself why the original submitted PR was not able to work properly on some platforms in the first place.

paglias commented:
> @paglias Sorry, tested the previous code on mac and seemed to work properly there but did not realize that for some reason it's not working on linux when testing the workflow more than once.

Actually I had tested on Mac as well but using Edge (which should be pretty similar to Chrome), @SabreCat have you ever seen something similar with modals?

paglias reviewed (commented):
sorry for the delay, I tested it and it works but I think `this.$root.$emit('habitica::dismiss-modal', 'members-modal'); ` is enough since it also calls `this.$root.$emit('bv:hide::modal', 'members-modal'); `

tsukimi2 commented:
> sorry for the delay, I tested it and it works but I think `this.$root.$emit('habitica::dismiss-modal', 'members-modal'); ` is enough since it also calls `this.$root.$emit('bv:hide::modal', 'members-modal'); `

@paglias Yes, I was aware that habitica::dismiss-modal calls "this.$root.$emit('bv:hide::modal".  However, when I tested it on mac, the members modal would still be displayed when I used only the line "this.$root.$emit('habitica::dismiss-modal', 'members-modal')".  When I tried to step over the code while the code was running, habitica::dismiss-modal did indeed seem to have called the line "this.$root.$emit('bv:hide::modal" within it, but for reason I do not yet understand it was not able to hide the members modal from there.  This happened on my mac running on the following:
Dev env: docker
Chrome: (version 80.0)
safari (v. 13.1)
edge (v.80.0)

In fact, on my mac, even if I just included the line "this.$root.$emit('bv:hide::modal" without including "this.$root.$emit('habitica::dismiss-modal", it would still work when tested against chrome, safari, and edge.  But on my ubuntu it's the other way around.

paglias commented:
Hmm but I tested on mac as well and got the opposite result... @negue do you have any idea ?

negue commented on website/client/src/components/userMenu/profile.vue:
```diff
@@ -5,16 +5,18 @@
+          <button
+            v-b-tooltip.hover.left="$t('sendMessage')"
+            class="btn btn-secondary message-icon"
+            @click="sendMessage()"
```
I think having a simple click executes the button click and the router-link which is then the reason the modals are opening up again.

negue replied:
`sendMessage` should be probably removed

paglias replied:
agreed with @negue , using the route itself seems to be enough

paglias replied:
So I think the button and the `sendMessage ` method can be removed

negue commented on website/client/src/pages/private-messages.vue:
```diff
@@ -630,8 +648,13 @@ export default {
-    const data = this.$store.state.privateMessageOptions;
+    // close members modal if the Private Messages page is opened in an existing tab
+    if (this.hasPrivateMessageOptionsOnPageLoad) {
+      this.$root.$emit('habitica::dismiss-modal', 'members-modal'); // hide members modal on linux
```
this would be needed once you also update the `membersModal.vue` :P 

but for now `profile` would be the better solution

paglias replied:
@negue are you suggesting to remove the line above or to simply change it to `this.$root.$emit('habitica::dismiss-modal', 'profile');`?

I'm also wondering if the line below is actually needed

negue reviewed (changes requested):
OK, after I wrote some information and tested it out github didn't save my message ... so rewriting some of it again.

I tested with Mac Vivaldi / Chrome and Windows Firefox, the current way worked (even *without* the additional `bc:hide::modal` call).

When you click the button/link (see comments) you will be redirected to the PM page, BUT pressing browser back, the modal repears again. I think this is because the modals are in our history-stack. Maybe we need something like https://router.vuejs.org/api/#replace to move to the PM page, but then the users won't be able to come back (with the browser back issue they can't anyway)

tsukimi2 committed:
Merge remote-tracking branch 'upstream/develop' into bugfix-pm-obscured

tsukimi2 committed:
Use replace in router-link to as per suggestion

tsukimi2 committed:
Suggestion by lint

tsukimi2 committed:
Remove sendMessage function in profile.vue as suggested in PR review

tsukimi2 committed:
Merge remote-tracking branch 'upstream/develop' into bugfix-pm-obscured

paglias commented:
thanks @tsukimi2 ! it's working well now, noted towards your 3rd tier
